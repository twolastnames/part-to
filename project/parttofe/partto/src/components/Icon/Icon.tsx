import React, { ForwardRefExoticComponent, RefAttributes } from "react";
import { rem } from "@mantine/core";
import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
  IconPlayerTrackNext,
  IconPlayerPlay,
  IconList,
  IconFile,
  IconCaretDown,
  IconCaretUp,
  IconChefHat,
  IconCooker,
  IconToolsKitchen,
} from "@tabler/icons-react";

import classes from "./Icon.module.scss";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}

export function Icon({ onClick, definition, size }: IconProps) {
  return React.createElement(definition, {
    ...(onClick
      ? {
          onClick,
          className: classes.clickable,
        }
      : {}),
    style: {
      width: rem((size || Size.Medium) as number),
      height: rem((size || Size.Medium) as number),
    },
  });
}

export const Next = IconPlayerTrackNext;
export const Start = IconPlayerPlay;
export const List = IconList;
export const File = IconFile;
export const Up = IconCaretUp;
export const Down = IconCaretDown;
export const ChefHat = IconChefHat;
export const Oven = IconCooker;
export const KitchenTools = IconToolsKitchen;
