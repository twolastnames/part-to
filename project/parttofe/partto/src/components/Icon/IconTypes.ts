import {
  Icon as TablerIcon,
  IconProps as TablerIconProps,
} from "@tabler/icons-react";
import { ForwardRefExoticComponent, RefAttributes } from "react";

export type IconType = ForwardRefExoticComponent<
  TablerIconProps & RefAttributes<TablerIcon>
>;

export enum Size {
  Small = 20,
  Medium = 30,
  Large = 50,
  ExtraLarge = 80,
}

export interface IconProps {
  definition: IconType;
  size?: Size;
  onClick?: () => void;
}
