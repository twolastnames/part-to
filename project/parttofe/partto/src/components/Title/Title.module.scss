@import "../../App.scss";

.title {
  display: flex;
  gap: var(--gap-medium);
  font: var(--font-strong-medium);
}

.icon {
  animation: popout var(--animation-tempo);
}

.text {
  animation: fadein var(--animation-tempo);
}
