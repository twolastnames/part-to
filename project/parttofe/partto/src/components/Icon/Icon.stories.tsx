import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Icon, Next } from "./Icon";
import { ShellProvider } from "../../providers/ShellProvider";
import { Size } from "./IconTypes";

const meta: Meta<typeof Icon> = {
  component: Icon,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Icon>;

export const Medium: Story = {
  args: { definition: Next },
};

export const Small: Story = {
  args: { definition: Next, size: Size.Small },
};

export const Large: Story = {
  args: { definition: Next, size: Size.Large },
};

export const ExtraLarge: Story = {
  args: { definition: Next, size: Size.ExtraLarge },
};
