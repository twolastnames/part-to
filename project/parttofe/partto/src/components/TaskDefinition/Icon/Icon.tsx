import React from "react";

import classes from "./Icon.module.scss";
import { ClassNames, IconProps } from "./IconTypes";
import {
  Duty as DutyIcon,
  Imminent as ImminentIcon,
  Icon as PhysicalIcon,
  Task as TaskIcon,
} from "../../Icon/Icon";

export const Duty: ClassNames = {
  imminent: classes.hidden,
  duty: classes.icon,
  task: classes.hidden,
};

export const Task: ClassNames = {
  imminent: classes.hidden,
  duty: classes.hidden,
  task: classes.icon,
};

export const Imminent: ClassNames = {
  imminent: classes.icon,
  duty: classes.hidden,
  task: classes.hidden,
};

export function Icon({ classNames: { task, duty, imminent } }: IconProps) {
  return (
    <>
      <span className={imminent}>
        <PhysicalIcon definition={ImminentIcon} />
      </span>
      <span className={duty}>
        <PhysicalIcon definition={DutyIcon} />
      </span>
      <span className={task}>
        <PhysicalIcon definition={TaskIcon} />
      </span>
    </>
  );
}
