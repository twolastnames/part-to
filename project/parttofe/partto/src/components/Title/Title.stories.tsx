import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Title } from "./Title";
import { ShellProvider } from "../../providers/ShellProvider";

const meta: Meta<typeof Title> = {
  component: Title,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Title>;

export const Simple: Story = {
  args: {},
};
