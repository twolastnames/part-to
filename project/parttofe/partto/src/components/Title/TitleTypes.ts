import { PropsWithChildren } from "react";
import { IconType } from "../Icon/IconTypes";

export interface TitleProps extends PropsWithChildren {
  icon: IconType;
}
