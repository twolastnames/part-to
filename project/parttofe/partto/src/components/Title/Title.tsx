import React from "react";

import classes from "./Title.module.scss";
import { TitleProps } from "./TitleTypes";
import { Icon } from "../Icon/Icon";
import { Size } from "../Icon/IconTypes";

export function Title({ icon, children }: TitleProps) {
  return (
    <div className={classes.title} data-testid="Title">
      <div className={classes.icon}>
        <Icon definition={icon} size={Size.Small} />
      </div>
      <div className={classes.text}>{children}</div>
    </div>
  );
}
