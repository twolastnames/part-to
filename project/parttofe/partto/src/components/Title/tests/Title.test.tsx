import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { Title } from "../Title";
import { ChefHat } from "../../Icon/Icon";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Title icon={ChefHat}>Chef hat for no reason</Title>
    </ShellProvider>,
  );
  const component = screen.getByTestId("Title");
  expect(component).toMatchSnapshot();
});
