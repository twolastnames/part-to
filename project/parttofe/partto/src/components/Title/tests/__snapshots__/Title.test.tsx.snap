// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="title"
  data-testid="Title"
>
  <div
    class="icon"
  >
    <svg
      class="tabler-icon tabler-icon-chef-hat "
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 3c1.918 0 3.52 1.35 3.91 3.151a4 4 0 0 1 2.09 7.723l0 7.126h-12v-7.126a4 4 0 1 1 2.092 -7.723a4 4 0 0 1 3.908 -3.151z"
      />
      <path
        d="M6.161 17.009l11.839 -.009"
      />
    </svg>
  </div>
  <div
    class="text"
  >
    Chef hat for no reason
  </div>
</div>
`;
